design-docs/junit-platform.md

```diff
@@ -0,0 +1,79 @@
+
+## Phase 0 - Identify Needed Gradle Abstractions
+
+Identify the abstractions and features needed to support arbitrary test frameworks with minimal coupling to their semantics.
+
+### Initial Thoughts
+
+* Core Infrastructure (JVM semantics may need to be extracted out if native or any other future languages/platforms should participate.)
+    * Identification - There needs to be a framework-agnostic test ID class (`TestDescriptor` exists now) for the below features
+    * Events - Current `TestListener` (really one of the internal protocols for this) seems to be the glue that feeds the other items below. This seems helpful, but needs to be usable for plugins in the public API.
+    * Logging - Current `TestLoggingContainer` provides a way to customize the inline reporting of test results. This seems helpful to all test frameworks. There may be question of what level to provide this customization. Should it be a global extension or specific to each test task.
+    * Reporting - Test frameworks usually provide their own reporting capabilities, so it wouldn't be strictly necessary to handle this directly in Gradle. However, aggregation of reports can be beneficial as well as consistent formatting.
+    * Tooling API (by extension BuildShip) - It should be possible to execute, see the results of, rerun, run individual, etc for arbitrary test frameworks through the Tooling API. This should be reflected in consumers like BuildShip as well.
+* Test Platform/Framework
+    * Either through a base task or just exposure of the right hooks, a plugin should be able to support any arbitrary test framework and get all of the infrastructure benefits above.
+    * The detection and execution models of those test frameworks should not be assumed by the core infrastructure.
+
+## Phase 1 - Test Infrastructure Changes
+
+Enhance the test infrastructure to support the core features that Gradle cares about independent of test frameworks. Provide abstractions to allow test frameworks/platforms to be implemented through plugins (using only public APIs).
+
+*Details TBD*
+
+## Phase 2 - JUnit Platform Support
+
+Implement support for execution of tests through the JUnit Platform. This should leverage the test infrastructure improvements from phase 1. (May need to happen concurrently with phase 1.)
+
+*Details TBD*
+
+## Phase 3 - JUnit "Vintage" and TestNG Support
+
+Support for JUnit "Vintage" (3 and 4) and TestNG leveraging the test infrastructure improvements, through one or more of the following:
+
+* Adapting the `Test` task to leverage the new infrastructure
+* Implementing new plugins/tasks to execute those tests using the new infrastructure
+* Relying on the JUnit Platform to have a `TestEngine` that can execute them
+
+*Details TBD*
+
+## Phase 4 - Evaluate Native Support
```

Comment:
From what I understand JUnit 5 is a test platform on the JVM. We will definitely need to clarify if it could be used for native support as well.